src/common/auth/fetch_token.js

```diff
@@ -8,6 +8,9 @@ export default function fetchToken(clientId, code) {
     body: data,
   }).then((resp) => {
     if (!resp.ok) throw new Error('Unable to fetch tokens');
-    return resp.json();
+    const tokens = resp.json();
+    // shorten access token life 30 seconds to cover network cost
+    tokens.expires = ((tokens.expires_in - 30) * 1000) + Date.now();
```

Comment:
Let's not shorten it 30 seconds here, it's up to the pace that uses it to decide this.

Also, please pass the value to the constructor of `Date.now()` instead of depending on type conversion.

Reply:
> Also, please pass the value to the constructor of Date.now() instead of depending on type conversion.

I didn't understand this point, Date.now() returns number of milliseconds since 1 January, 1970, 00:00:00, UTC. IMO it is perfect to store it for comparison usage, we don't need `Date` object here.

Reply:
> Let's not shorten it 30 seconds here

The idea for this change is to avoid unnecessary invalid auth error, If we don't shorten it, it is highly possible the server will expires token earlier than client. If you feel 30 seconds is too long, we can reduce it to 10 or 5.

Reply:
Oh, I didn't realize `Date.now()` returns an integer. My bad.

It is up to the place that checks for expired to decide the offset. So this 30 seconds adjustment should happen inside `home-assistant-js-websocket`.